Crowdfund parachain deposit

gavofyork opened the pull request.

shawntabrizi commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,539 @@
+				// Now in end period; record it
+				fund.last_contribution = Some(now);
+				if let Some(c) = fund.last_contribution {
+					if c != now {
```
This will always be false since we just set `last_contribution` to `Some(now)`

What is the intended logic here?

shawntabrizi replied:
Maybe you intend to update the value _after_ the check is made.

gavofyork replied:
you're right, yes

shawntabrizi commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,539 @@
+		
+		/// Note that a successful fund has lost its parachain slot, and place it into retirement.
+		fn begin_retirement(_origin, #[compact] index: FundIndex) {
+			// origin unimportant.
```
When you say "origin unimportant", I think you mean "allow for any origin".

But I think we should explicitly check that it is **not** inherent origin, else a malicious actor could spam the chain with function calls, causing storage lookup and other logic.

What do you think?

gavofyork replied:
sure we can `ensure_signed`.

shawntabrizi replied:
I have updated all functions which did not use origin to have `ensure_signed`.

shawntabrizi commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,539 @@
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			let _parachain_id = fund.parachain.take().ok_or("fund has no parachain")?;
+			let account = Self::fund_account_id(index);
+			ensure!(T::Currency::free_balance(&account) >= fund.raised, "funds not yet returned");
```
This seems to be the only "check" that this function can be called to end a crowdfund.
However, it is not clear yet that someone will not be able to call this:
* When the crowdfund has collected all the funds
* Before `handle_bid` is called on the fund, reserving them

It seems we imply that a fund having a `para_id` means that it has completed onboarding, but that does not seem to be the full story...

shawntabrizi replied:
Having finished reading the full code and understanding to the best of my ability, I think this comment still stands.

gavofyork replied:
> It seems we imply that a fund having a para_id means that it has completed onboarding

correct. what's the rest of the story?

shawntabrizi replied:
I think I misunderstood, and it does seem right. The story:

* Parachain opens crowdfund, collects as much funds as possible up to cap. No parachain id assigned to it.
* When auction `is_ending` a bid is sent from the crowdfund with the total amount raised so far, reducing its free balance.
* A crowdfund can finally win the bid, and someone will call `onboard` which will ultimately assign it a para_id, but only if the onboarding process succeeds in `slots` too.

At no point will a fund have both its total funds raised and a para_id until after it has lost its slot.

Thanks!

shawntabrizi commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,539 @@
+			ensure!(value >= T::MinContribution::get(), "contribution too small");
+
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			fund.raised += value;
```
No reason not to do a checked add here.

gavofyork replied:
no reason to do one, either.

gavofyork replied:
(but if you want to do it, that's ok)

shawntabrizi commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,539 @@
+
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			fund.raised += value;
+			ensure!(fund.raised <= fund.cap, "contributions exceed cap");
```
It seems a fund can be trapped between two requirements:
* Contribution must make the total raised to be less than or equal to the fund cap (this)
* Contribution must be more than the min contribution

It seems like we should do something like:
* Before contributing, make sure that the total fund raised at the time is less than the fund cap
* Make sure value contributing is more than min contribution
* Only add the amount of funds needed to make it reach the fund cap, even if it is less than min contribution

In this case, there may be one contributor with an amount smaller than the min contribution, but only one, which means that the rule below is only broken one time. Probably not a big deal at that point.

```
//! Contributions must be of at least `MinContribution` (to account for the resources taken in
//! tracking contributions)
```

gavofyork replied:
who care if it can't quite 100% optimise its cap by a piece of dust?!

this is a complete non-issue. leave it alone.

shawntabrizi commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,539 @@
+		#[weight = TransactionWeight::Basic(100_000, 10)]
+		fn create(
+			origin,
+			#[compact] end: T::BlockNumber,
```
Missing check that `end` is greater than `now`.

Should there also be some "minimum" crowdfund time? Would never make sense to have an `end` be one block after `now`.

gavofyork replied:
i don't think a prescribing a minimum is necessary: it doesn't provide any additional economic protections and logic to save the users from themselves should be in the UI, not increasing bug surface of the runtime.

shawntabrizi commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,539 @@
+			#[compact] end: T::BlockNumber,
+			#[compact] cap: BalanceOf<T>,
+			#[compact] first_slot: T::BlockNumber,
+			#[compact] last_slot: T::BlockNumber
```
Missing check that `last_slot` is great then `first_slot`.

This is important to make `handle_bid` infallible.

shawntabrizi replied:
Could be resolved by simply calling `new_bounded` somewhere and checking it returns ok

gavofyork replied:
> Missing check that last_slot is great then first_slot.

sure

> This is important to make handle_bid infallible

`handle_bid` doesn't need to be infallible.

shawntabrizi commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,539 @@
+			origin,
+			#[compact] end: T::BlockNumber,
+			#[compact] cap: BalanceOf<T>,
+			#[compact] first_slot: T::BlockNumber,
```
Missing check that `first_slot` is greater than `end`, but to really make sure `handle_bid` is infallible, we need to check that `first_slot` is greater than `first_lease_period` so that `new_bounded` in slot range will not fail.

Seems like we should call `new_bounded` somewhere in `create` to make sure that it will not fail later.

shawntabrizi replied:
I suggest that we do not let the user input `end`, and rather we automatically set end to:

```
// Last possible block of an auction
let end = early_end + T::EndingPeriod::get();
```

for the current auction. In addition, we will check that the auction `is_in_progress` for things like contributions.

shawntabrizi replied:
I have done this in `Automatically assign end for crowdfund`, which we can revert if you don't like the idea.

gavofyork replied:
the current slot might change before the auction succeeds, so `handle_bid` can always be fallible at present. so far the assumption is that nobody would want to bid on an auction that they couldn't use. maybe that's invalid. but calling `new_bounded` here won't help other than checking that `first_slot < last_slot <= first_slot + 3`.

shawntabrizi commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,539 @@
+							/// ever changes, then some sort of conversion will be needed here.
+							sub: 0,
+						});
+						// TODO for review: Can this fail? Anything we can do to handle possible errors?
```
Some comments above fix other possible error conditions. One more place it can fail is:

```
			// This must overlap with all existing ranges that we're winning on or it's invalid.
			ensure!(current_winning.iter()
				.enumerate()
				.all(|(i, x)| x.as_ref().map_or(true, |(w, _)|
					w != &bidder || range.intersects(i.try_into()
						.expect("array has SLOT_RANGE_COUNT items; index never reaches that value; qed")
					)
				)),
				"bidder winning non-intersecting range"
			);
```

Should we test for this somewhere throughout this auction process? is it okay if this fails?

gavofyork replied:
it's not actually possible for it to fail in this way since each fund is a unique bidder. but yeah, it's not a big problem if it fails here in principle (except for the individuals involved, but they should have been more careful).

gavofyork commented on runtime/src/slots.rs:
```diff
@@ -636,7 +636,7 @@ impl<T: Trait> Module<T> {
 	/// - `last_slot`: The last lease period index of the range to be bid on (inclusive).
 	/// - `amount`: The total amount to be the bid for deposit over the range.
-	fn handle_bid(
+	pub fn handle_bid(
```
shouldn't be `pub`.

gavofyork replied:
and *really* should be in `slots.rs`!

4meta5 commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,547 @@
+//! @WARNING: For funds to be returned, it is imperative that this module's account is provided as
+//! the offboarding account for the slot. In the case that a parachain supplemented these funds in
+//! order to win a later auction, then it is the parachain's duty to ensure that the right amount of
+//! funds ultimately end up in module's fund sub-account. If the funds do not arrive, then
```
```suggestion
//! funds ultimately end up in module's fund sub-account.
```

I assume the hanging last sentence might have read: "If the funds do not arrive, then the parachain will not be able to bid on future auctions with the funds." This sentence would be redundant given the previous sentence.

4meta5 commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,547 @@
+	/// least ExistentialDeposit.
+	type MinContribution: Get<BalanceOf<Self>>;
+
+	/// The period of time (in blocks) between after an unsuccessful crowdfund ending where
```
```suggestion
	/// The period of time (in blocks) after an unsuccessful crowdfund ending when
```

4meta5 commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,547 @@
+	/// The total amount raised.
+	raised: Balance,
+	/// Block number after which the funding must have succeeded. If not successful at this number
+	/// the everyone may withdraw their funds.
```
```suggestion
	/// then everyone may withdraw their funds.
```

4meta5 commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,955 @@
+		buf.extend_from_slice(b"crowdfund");
+		buf.extend_from_slice(&index.to_le_bytes()[..]);
+
+		CHILD_STORAGE_KEY_PREFIX.iter()
```
```suggestion
		CHILD_STORAGE_KEY_PREFIX.into_iter()
```

shawntabrizi replied:
Where is this suggestion coming from? What is the difference?

4meta5 replied:
`.iter()` clones the object which isn't necessary because you're not using `CHILD_STORAGE_KEY_PREFIX` again so you might as well use `.into_iter()`, unless I'm missing something

4meta5 replied:
@shawntabrizi  For clarification, `into_iter` is preferrable in the scenario when you can consume the iterator because the object is not used thereafter. If there was an encoding requirement, `iter` would be preferred, but there is no compilation error so this is not the case.

4meta5 commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,955 @@
+
+		CHILD_STORAGE_KEY_PREFIX.iter()
+			.chain(b"default:")
+			.chain(T::Hashing::hash(&buf[..]).as_ref().iter())
```
```suggestion
			.chain(T::Hashing::hash(&buf[..]).as_ref().into_iter())
```

shawntabrizi commented:
@gavofyork This is ready for review.

[Here is a diff](https://github.com/paritytech/polkadot/compare/7ee6b272964e29501299d8a24eaa12b0472b454a..79a1c8d0df283742ff1831bbfb215dcb006d4e9a) from where you left off and where I am now. It is a bit noisy since I had to merge master once.

I will summarize the core parts which you should double check where significant changes were made:

* I removed the `Currency` trait you defined, and instead inherited `Currency` from `slots::Trait`
* I removed the `WasEnding` boolean storage item, as I did not need it for any of the logic
* I introduced an event for every dispatchable function
* I removed `end` as an input when creating a crowdfund. The `end` block for a crowdfund is automatically set to be the last possible block for the current auction. (`let end = early_end + T::EndingPeriod::get();`)
* I have `FundCount` start at 0, whereas it was starting at 1 before. So the first crowdfund will have index 0.
* I modified the logic around `last_contribution` and `NewRaise`. You had previously set some magic around `last_contribution` being set to `Some()`, and then it would picked up by `NewRaise`, and well... the intention made sense, but implementation did not. I have done the following:
    * `last_contribution` set to `None` to start
    * **If** first contribution is detected (`is_none`) we add the fund to `NewRaise`
    * **Else**  if contribution happens during the ending period, and we did not already add it to `NewRaise` this block (multiple contributions on the same block), we add it to `NewRaise`
    * At the end, `last_contribution` is updated to the current block, and always does represent the actual last contribution.
* `OnFinalize` logic updated slightly for this `NewRaise` usage
* Functions where `origin` was not used, I instead added `ensure_signed(origin)`

Final question I had when looking through the code:

* Should there be an upper bound to the number of crowdfunds per auction, or `NewRaise` per block? We do an `on_finalize` loop/iteration where we have no such bounds. The first block when an auction `is_ending` will need to loop through all crowdfunds which have received funds to that point and submit a bid. It makes sense to me to set a limit to number of crowdfunds, which is checked during `create()`.

xlc commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,1104 @@
+		MODULE_ID.into_sub_account(index)
+	}
+
+	pub fn id_from_index(index: FundIndex) -> Vec<u8> {
```
why not using `double_map`?

shawntabrizi replied:
Gav's comment in the docs was:

```
//! Contributing funds is permissionless. Each fund has a child-trie which stores all
//! contributors account IDs together with the amount they contributed; the root of this can then be
//! used by the parachain to allow contributors to prove that they made some particular contribution
//! to the project (e.g. to be rewarded through some token or badge). The trie is retained for later
//! (efficient) redistribution back to the contributors.
```

I too had this question, but when talking with @thiolliere, he mentioned that this would isolate the crowdfund tracking trie from the shared module storage trie. Thus, the root of the trie would only represent contributions to that crowdfund. Will let @gavofyork complete the story

xlc replied:
I see. It will be very useful if this can be generalized later in Substrate as I can see many different usages for this pattern.

shawntabrizi replied:
I think this PR is supposed to help generalize the usage of child tries: https://github.com/paritytech/substrate/pull/2209

Maybe you mean something more though?

xlc replied:
it is a start point, but not a complete developer friendly solution. e.g. no metadata integration and hence hard JS SDK integration.
I am looking for something like
```
decl_child_storage! {
  ParentKey: FundIndex;
  Contribution: map AccountId => Balances;
}
```

thiolliere replied:
yes having child storage stuff into metadata seems important, I'm not aware of any work on that currently, but I think it is important as well.
The PR 2209 should just change a bit the usage of the child trie.

xlc replied:
https://github.com/paritytech/substrate/issues/3187

gavofyork commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,1104 @@
+
+decl_event!(
+	pub enum Event<T>
+		where
```
move on to previous line

gavofyork commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,1104 @@
+decl_event!(
+	pub enum Event<T>
+		where
+	<T as system::Trait>::AccountId,
```
add an indent

gavofyork commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,1104 @@
+		/// Create a new crowdfunding campaign for a parachain slot deposit for the current auction.
+		#[weight = TransactionWeight::Basic(100_000, 10)]
+		fn create(
+			origin,
```
i typically leave `self` and `origin` on the first line as they're exceptional (non-typed) arguments.

gavofyork commented:
> * I removed the `Currency` trait you defined, and instead inherited `Currency` from `slots::Trait`

ok.

> * I removed the `WasEnding` boolean storage item, as I did not need it for any of the logic

seems ok.

> * I introduced an event for every dispatchable function

probably ok.

> * I removed `end` as an input when creating a crowdfund. The `end` block for a crowdfund is automatically set to be the last possible block for the current auction. (`let end = early_end + T::EndingPeriod::get();`)

incorrect - a crowdfund can span multiple auctions.

> * I have `FundCount` start at 0, whereas it was starting at 1 before. So the first crowdfund will have index 0.

probably fine. you did add an overflow check for a u32. for it to overflow, it would require over 4,000,000,000 transactions to be sent, each paying a fee and making a deposit. Assuming the fee and deposit combination cost for a crowdfund is $1,000 and the weight is such that we can start up to 10 crowdfunds per block, then you'd need to be both a multi-trillionaire *and* wait around 76 years to execute a panic attack (which would be trivially patchable). i'm not convinced the extra 2 lines of code (and thus attack surface) is worth it for something than can be proven to be quite so improbable.

my general feeling is that runtime correctness checks should only be conducted when there's a reasonable (if unlikely) possibility of them be triggered. in cases where it's demonstrably impossible (for any sensible state of the world) then i'm against littering the code since.

> * I modified the logic around `last_contribution` and `NewRaise`. You had previously set some magic around `last_contribution` being set to `Some()`, and then it would picked up by `NewRaise`, and well... the intention made sense, but implementation did not. I have done the following:
>   
>   * `last_contribution` set to `None` to start
>   * **If** first contribution is detected (`is_none`) we add the fund to `NewRaise`
>   * **Else**  if contribution happens during the ending period, and we did not already add it to `NewRaise` this block (multiple contributions on the same block), we add it to `NewRaise`
>   * At the end, `last_contribution` is updated to the current block, and always does represent the actual last contribution.
> * `OnFinalize` logic updated slightly for this `NewRaise` usage

this will need full inspection as i spent some time thinking about and optimising the logic.

does it also do *exactly* the right thing when the fund happens over multiple auctions? 

> * Functions where `origin` was not used, I instead added `ensure_signed(origin)`

ok.

> * Should there be an upper bound to the number of crowdfunds per auction, or `NewRaise` per block? We do an `on_finalize` loop/iteration where we have no such bounds. The first block when an auction `is_ending` will need to loop through all crowdfunds which have received funds to that point and submit a bid. It makes sense to me to set a limit to number of crowdfunds, which is checked during `create()`.

a limit is probably fine to introduce as a sanity check, but the costs involved in starting a funding campaign should be sufficient to avoid that limit ever being reached.

gavofyork commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,1110 @@
+				cap: cap,
+				last_contribution: None,
+				first_slot: first_slot,
+				last_slot: last_slot,
```
these `name: name,` lines can be omitted with a `..` at the end. less clutter -> good.

gavofyork commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,1110 @@
+			let (_, early_end) = <slots::Module<T>>::auction_info().ok_or("no auction in progress")?;
+
+			// End of the crowdfund will be the last possible block for the ongoing auction
+			let end = early_end + T::EndingPeriod::get();
```
incorrect

gavofyork commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,1110 @@
+			ensure!(first_slot < last_slot, "last slot must be greater than first slot");
+			ensure!(last_slot <= first_slot + 3.into(), "last slot cannot be more then 3 more than first slot");
+			// Check an auction is in progress, and extract the `early_end` block
+			let (_, early_end) = <slots::Module<T>>::auction_info().ok_or("no auction in progress")?;
```
an auction need not be in progress for a fund to be started.

gavofyork commented on runtime/src/crowdfund.rs:
```diff
@@ -0,0 +1,1110 @@
+//! a deposit must be paid of the amount `SubmissionDeposit`. Substantial resources are taken on
+//! the main trie in tracking a fund and this accounts for that.
+//!
+//! Funds may be set up during an auction period; their closing time is fixed at creation (as a
```
It is "may", not "must". Funds may last for many auctions.

gavofyork commented:
@shawntabrizi would be good to get tests in for funds that span multiple auctions and funds that begin prior to any auction.

shawntabrizi commented:
> @shawntabrizi would be good to get tests in for funds that span multiple auctions and funds that begin prior to any auction.

Yes, I can do this asap.

shawntabrizi commented:
> my general feeling is that runtime correctness checks should only be conducted when there's no possibility of making a static correctness check. in cases where it's demonstrably impossible or impractical for an error to be triggered then i'm against littering the code.

This makes sense to me, but now I struggle to understand how best to teach a concept like this to new runtime developers. Will follow up with you offline.

> does it also do exactly the right thing when the fund happens over multiple auctions?

You are right it does not work over multiple auctions. The updated logic you introduced makes sense to me, and handles this situation well. I will write tests to check it.

shawntabrizi commented:
@gavofyork a crowdfund with an end date longer than a single auction seems to work with the last test I added. However, from one action end to the next auction, there **must** be a contribution to the fund in order for it to be added to `NewRaise` again, and submit a bid.

Is this okay?

gavofyork commented:
The decision on when/whether to place an item in `NewRaise` should be tested thoroughly... Does it check when the fund has a bid outside of an ending period, then in a later auction, also outside the ending period, a winning bid?

gavofyork commented:
ok - i'm reasonably happy with it.

gavofyork commented:
@shawntabrizi once you're happy with it, give it a final review and we can merge.

shawntabrizi commented:
Note: Right now only `create` has a transaction weight. Could need a follow up after doing analysis on the functions.
